### PR TITLE
Add evaluate to trainer

### PR DESCRIPTION
**Pull request by Het-Shah**
Added Evaluate to classical RL trainer. Thoughts on this?

### CODE REVIEW

**Comment by Sharad24** [comment]
Still working on this?

**Comment by Het-Shah** [comment]
Is this fine or should I add something else?

**Comment by Sharad24** [comment]
Looks good to me, have you tested this on all algos?

**Comment by Sharad24** [comment]
Maybe we should think about renaming Trainer to Runner in the future, if we want to have functions like evaluate.

**Comment by Het-Shah** [comment]
> Yup i tested this on both QLearning and SARSA.

**Comment by Het-Shah** [comment]
> Maybe we should think about renaming Trainer to Runner in the future, if we want to have functions like evaluate.

Seems like a better idea to me because right now for evaluating we need to run `agent.evaluate`, which seems odd to me

**Comment by Het-Shah** [comment]
Should I add evaluate to trainer in deepRL as well in this PR only?

**Comment by Het-Shah** [comment]
I am going to add `evaluate` for DeepRL trainers as well.

**Comment by Sharad24** [comment]
Wow. Sounds good!

**Comment by Het-Shah** [comment]
I have tested every deeprl agent on CartPole-v0, got a mean reward of >120 for every agent after 2 epochs. Let me know if you are not able to replicate.

**Comment by Sharad24** [comment]
Ok cool. What about DDPG, TD3 and SAC?

**Comment by Het-Shah** [comment]
> DDPG, TD3 and SAC?

Yup these too, made some changes to them.

**Comment by Het-Shah** [comment]
Oh wait SAC is left. I'll add tests as well. Sorry.

**Comment by Sharad24** [comment]
How but? DDPG, TD3 and SAC are not supposed to work for discrete envs 🤣

**Comment by Sharad24** [comment]
Maybe check Pendulum-v0 for them

**Comment by Sharad24** [comment]
Can you save the agents for these? And record a video for them? (We could add these in readme)

Recording just requires [ monitoring the env](https://stackoverflow.com/questions/52636899/python-openai-gym-monitor-creates-json-files-in-the-recording-directory).

**Comment by Het-Shah** [comment]
> How but? DDPG, TD3 and SAC are not supposed to work for discrete envs 

I don't know. It ran successfully :P

**Comment by Sharad24** [comment]
Ok. I think CartPole-v0 should reach 200, as its the easiest environment. 

I dont know how but this was an error with Vecenvs as well (It was only reaching a reward of 120). We should probably debug this.

Pendulum-v0 reward should reach around -120
CartPole-v0 should be 200
CartPole-v1 should be 500

**Comment by Het-Shah** [comment]
> I think CartPole-v0 should reach 200

Yup i tested this for DQN without vectored env it reaches 200. 

> I dont know how but this was an error with Vecenvs as well 

Regarding this I'll check the subprocess code again, maybe there is an issue there. 

> Can you save the agents for these? And record a video for them?

Cool, I'll add these in another PR. Let's not add it here.

**Comment by Sharad24** [comment]
Ok sure, sounds good to me. But just save the agents and video (or hyperparams) to avoid doing the hardworking

**Comment by Het-Shah** [comment]
Yea sure.

**Comment by Het-Shah** [comment]
Also, why are there buffer errors now. I did not make any change to them.

**Comment by Sharad24** [comment]
Not sure, you could try running the tests locally?

**Comment by Het-Shah** [comment]
Yea now even locally they are showing up. My ubuntu really sucks now.

File: genrl/deep/agents/ddpg/ddpg.py [hunk 1]
```diff
@@ -128,16 +128,17 @@ def __init__(
         self.create_model()
 
     def create_model(self):
-        state_dim = self.env.observation_space.shape[0]
-        action_dim = self.env.action_space.shape[0]
+        # state_dim = self.env.observation_space.shape[0]
+        # action_dim = self.env.action_space.shape[0]
+        state_dim, action_dim, self.discrete = self.get_env_properties()
```
Comment (Sharad24) [inline genrl/deep/agents/ddpg/ddpg.py:133]: Not related to build errors

Discrete should always be False here (DDPG, TD3, SAC). So maybe we should add a condition to check if self.discrete is True. If not, raise error? What do you think?

Comment (Het-Shah) [reply genrl/deep/agents/ddpg/ddpg.py:133]: yea, correct this is why CartPole was working :(

**Comment by Het-Shah** [comment]
will squash commits this time :P

**Comment by Sharad24** [comment]
Travis didnt fail this time? What changed?

**Comment by Het-Shah** [comment]
> I don't know, I restarted everything now. Pushing every agent one by one.

**Comment by Het-Shah** [comment]
I am done now I guess with every agent. Review this lol.

File: genrl/deep/agents/ddpg/ddpg.py [hunk 2]
```diff
@@ -128,8 +128,9 @@ def __init__(
         self.create_model()
 
     def create_model(self):
-        state_dim = self.env.observation_space.shape[0]
-        action_dim = self.env.action_space.shape[0]
+        state_dim, action_dim, discrete = self.get_env_properties()
+        if discrete == True:
+            raise NotImplementedError
```
Comment (Sharad24) [inline genrl/deep/agents/ddpg/ddpg.py:133]: Can you add a message to this error?

File: genrl/deep/agents/vpg/vpg.py [hunk 3]
```diff
@@ -116,19 +116,19 @@ def create_model(self):
...
+        # action_lim = None
+        # if isinstance(self.env.action_space, gym.spaces.Discrete):
+        #     action_dim = self.env.action_space.n
+        #     discrete = True
+        # elif isinstance(self.env.action_space, gym.spaces.Box):
+        #     action_dim = self.env.action_space.shape[0]
+        #     action_lim = self.env.action_space.high[0]
+        #     discrete = False
+        # else:
+        #     raise NotImplementedError
```
Comment (Sharad24) [inline genrl/deep/agents/vpg/vpg.py:130]: Can remove this now, I think

File: genrl/deep/agents/ddpg/ddpg.py (hunk 2 above)
Comment (Het-Shah) [reply genrl/deep/agents/ddpg/ddpg.py:133]: Ok, I am changing the message to "Discrete environments not supported".

File: genrl/deep/agents/vpg/vpg.py (hunk 3 above)
Comment (Het-Shah) [reply genrl/deep/agents/vpg/vpg.py:130]: Yes sorry.

File: genrl/deep/agents/ddpg/ddpg.py (hunk 2 above)
Comment (Sharad24) [reply genrl/deep/agents/ddpg/ddpg.py:133]: "...not supported for {}".format(algo_name)

**Comment by Sharad24** [comment]
I think changes are minimal so we can merge. Probably does not need another review from @ajaysub110

**Comment by Het-Shah** [comment]
Totally upto you.

File: genrl/deep/agents/ddpg/ddpg.py (hunk 2 above)
Comment (Het-Shah) [reply genrl/deep/agents/ddpg/ddpg.py]: oh sorry okay.

**Comment by Het-Shah** [comment]
Aah more work. Nice :P

**Comment by ajaysub110** [comment]
I'll fix them

**Comment by Het-Shah** [comment]
wait i pushed too lol

**Comment by Het-Shah** [comment]
oh i am cancelling the push lol

**Comment by ajaysub110** [comment]
Lol. Sorry boss. I invaded your PR

**Comment by Het-Shah** [comment]
Sorry from my side. You had to do it :(

**Pull request merged**